test/NuGet.Core.Tests/NuGet.Packaging.Test/ContentModelTests/PatternExpressionMatchTests.cs

```diff
@@ -0,0 +1,185 @@
+            Assert.NotNull(result);
+            Assert.Equal(path, result.Path);
+        }
+
+        [Fact]
+        public void Match_LiteralSegment_ShouldNotMatch()
+        {
+            // should not match the literal segment pattern because file.txt != file2.txt
+            Dictionary<string, ContentPropertyDefinition> propertyDefinitions = new()
+            {
+                { "type", new ContentPropertyDefinition("text") }
+            };
+            PatternDefinition pattern = new PatternDefinition("content/file.txt");
+            PatternExpression expression = new PatternExpression(pattern);
+            var path = "content/file2.txt";
+            ContentItem result = expression.Match(path, propertyDefinitions);
+            Assert.Null(result);
+        }
+
+        [Fact]
+        public void Match_LiteralSegmentAndTokenSegment_ShouldMatch()
+        {
+            //We have a parser for the token "name" which takes in any type of string and returns it.
+            //Therefore file.txt should be added to the result content item
+            Dictionary<string, ContentPropertyDefinition> propertyDefinitions = new()
+            {
+                { "name", new ContentPropertyDefinition("any", parser: (o, t) => o) }
+            };
+            //this pattern has both a token segment and a literal segment
+            PatternDefinition pattern = new PatternDefinition("content/{name}");
+            PatternExpression expression = new PatternExpression(pattern);
+            var path = "content/file.txt";
+            ContentItem matchResult = expression.Match(path, propertyDefinitions);
+            Assert.True(matchResult.Properties.TryGetValue("name", out object result));
+            Assert.Equal("file.txt", result);
+        }
+
+        [Fact]
+        public void Match_LiteralSegmentAndTokenSegment_ShouldNotMatch()
+        {
+            // We do not have a parser for the token segment name.
+            //As a result it should not be able to match and return a content item
+            Dictionary<string, ContentPropertyDefinition> propertyDefinitions = new()
+            {
+                { "name", new ContentPropertyDefinition("any") }
+            };
+            // this pattern has both a token segment and a literal segment
+            PatternDefinition pattern = new PatternDefinition("content/{name}");
+            PatternExpression expression = new PatternExpression(pattern);
+            var path = "content/file.txt";
+            ContentItem matchResult = expression.Match(path, propertyDefinitions);
+            Assert.Null(matchResult);
+        }
+
+        [Fact]
+        public void Match_TokenSegmentMatchOnly_ShouldNotAddProperty()
+        {
+            // We have a parser that would match any field however the field "file.txt" here should not be added to the key "name" in the content Item
+            // because we used ? in the pattern for the token name and as a result the token is a matchonly token.
```

Comment:
What does "matchonly token" mean?

I'm not saying this is wrong. The test is passing, so clearly it's how the code works. But I find it confusing. When I look at the real patterns we use in our product code, `ManagedCodeConventions.cs`, we have a bunch of patterns that end in `{any?}`. Sometimes we use it like `contentFiles/{codeLanguage}/{tfm}/{any?}`, which I would expect to match both `contentFiles/any/any/file.txt` and also `contentFiles/any/any/subdirectory/file.txt`, but this test is asserting that it doesnt' match files in subdirectories?

edit: actually, it looks like the test is asserting that the expression matches the file, but the `{name}` "variable" didn't match anything. I think this comes back to my previous comment about "test the contract, not the implementation". I don't understand what user-experience this test is validating, it feels like an edge case that doesn't make sense.